### ISSUE TITLE
NuGet.exe commands are very slow when using CredentialProvider.VSS.exe

### ISSUE THREAD

**Issue by sanderaernouts** [NuGet/Home#7236]
## TL;DR;
We recently moved from a self-hosted NuGet server on-premise to VSTS package management. We now see a significant increase in the amount of time it takes to run commands through `nuget.exe`. This seems related to the `CredentialProvider.VSS.exe` not (correctly) caching the token that is generated to access the feeds. We use nuget.exe to download for example test data from our feeds so this increase from ~1 second to 5-10 seconds causes our these tests to be much much slower. A

## Details about Problem
**NuGet product used:** NuGet.exe
**NuGet version:** 4.7.1.5393
**OS version:** Microsoft Windows 10 Pro (10.0.17134 Build 17134)
**CredentialProvider.VSS.exe version:** 0.37.0
**Feed type**: VSTS package management

## Detailed repro steps so we can see the same problem

1. Make sure you have `CredentialProvider.VSS.exe` in `c:\users\<user>\AppData\Local\NuGet\CredentialProviders`
2. Run `nuget list` for a single package against a VSTS feed.
![image](https://user-images.githubusercontent.com/6184673/44567717-06fba480-a774-11e8-8fda-be8d658ca8b9.png)

This command takes between 5-10 seconds:
![image](https://user-images.githubusercontent.com/6184673/44567727-14b12a00-a774-11e8-889b-42228cb58e8d.png)

If you run the same command using a `NuGet.config` file that has a Personal Access Token configured for the same feed you get:
![image](https://user-images.githubusercontent.com/6184673/44567760-33172580-a774-11e8-8b85-4dcb7d77d758.png)

This command takes about 1 second:
![image](https://user-images.githubusercontent.com/6184673/44567764-37434300-a774-11e8-9dc4-7547acf838b8.png)

## Other suggested things
Looking at the verbose logs I notice a few things. First the credential provider fails to detect there already is a token, the provider seems to request a new token every time you call `Nuget.exe`. When it tries to do this, it fails to get the token with using Sps credential type and it falls back to Browser-based federated authentication and it successfully retrieves a token. Most of the time seems to be spend here. Which would be "ok" if the token was actually stored/cached somewhere and we wouldn't have to go trought this every time.

### Verbose Logs
NuGet Version: 4.7.1.5393
CredentialProvider.VSS: Microsoft Visual Studio Team Services Credential Provider for NuGet 0.37.0.0
CredentialProvider.VSS: Probing https://account.pkgs.visualstudio.com/_packaging/feed/nuget/v3/index.json
CredentialProvider.VSS: Probe response code: 401 Unauthorized
CredentialProvider.VSS: Credential type: Sps.
CredentialProvider.VSS: Has valid credentials: False.
CredentialProvider.VSS: Getting new credentials for source:https://account.pkgs.visualstudio.com/_packaging/feed/nuget/v3/index.json, scope:vso.packaging_write vso.drop_write
CredentialProvider.VSS: Getting authentication token from https://account.vssps.visualstudio.com/
CredentialProvider.VSS: Trying authentication provider Windows Integrated Authentication via Azure AD...
CredentialProvider.VSS: Authorization failure while retrieving a session token
CredentialProvider.VSS: Microsoft.VisualStudio.Services.Common.VssUnauthorizedException: VS30063: You are not authorized to access https://account.vssps.visualstudio.com.
CredentialProvider.VSS:    at Microsoft.VisualStudio.Services.Common.VssHttpMessageHandler.<SendAsync>d__17.MoveNext()
CredentialProvider.VSS: --- End of stack trace from previous location where exception was thrown ---
CredentialProvider.VSS:    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
CredentialProvider.VSS:    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
CredentialProvider.VSS:    at Microsoft.VisualStudio.Services.Common.VssHttpRetryMessageHandler.<SendAsync>d__3.MoveNext()
CredentialProvider.VSS: --- End of stack trace from previous location where exception was thrown ---
CredentialProvider.VSS:    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
CredentialProvider.VSS:    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
CredentialProvider.VSS:    at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.<SendAsync>d__46.MoveNext()
CredentialProvider.VSS: --- End of stack trace from previous location where exception was thrown ---
CredentialProvider.VSS:    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
CredentialProvider.VSS:    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
CredentialProvider.VSS:    at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.<SendAsync>d__43`1.MoveNext()
CredentialProvider.VSS: --- End of stack trace from previous location where exception was thrown ---
CredentialProvider.VSS:    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
CredentialProvider.VSS:    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
CredentialProvider.VSS:    at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.<SendAsync>d__27`1.MoveNext()
CredentialProvider.VSS: --- End of stack trace from previous location where exception was thrown ---
CredentialProvider.VSS:    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
CredentialProvider.VSS:    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
CredentialProvider.VSS:    at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.<SendAsync>d__26`1.MoveNext()
CredentialProvider.VSS: --- End of stack trace from previous location where exception was thrown ---
CredentialProvider.VSS:    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
CredentialProvider.VSS:    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
CredentialProvider.VSS:    at Microsoft.VisualStudio.Services.DelegatedAuthorization.Client.DelegatedAuthorizationHttpClient.<CreateSessionToken>d__11.MoveNext()
CredentialProvider.VSS: --- End of stack trace from previous location where exception was thrown ---
CredentialProvider.VSS:    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
CredentialProvider.VSS:    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
CredentialProvider.VSS:    at Microsoft.VisualStudio.Services.DelegatedAuthorization.Client.DelegatedAuthorizationHttpClient.<CreateSessionToken>d__8.MoveNext()
CredentialProvider.VSS: --- End of stack trace from previous location where exception was thrown ---
CredentialProvider.VSS:    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
CredentialProvider.VSS:    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
CredentialProvider.VSS:    at Microsoft.VisualStudio.Services.WebApi.TaskExtensions.SyncResult[T](Task`1 task)
CredentialProvider.VSS:    at VSS.NuGet.Authentication.SpsAuthTokenProvider.GetAuthToken(Uri spsUri, TimeSpan sessionLength, Boolean allowInteractive, String scope, Boolean discardExistingCredentials)
CredentialProvider.VSS: Trying authentication provider Browser-based federated authentication...
CredentialProvider.VSS: SPS CreateSessionToken Activity ID: 08662d9d-5357-4cac-bc40-2c93ce2e2ba0
CredentialProvider.VSS: CreateSessionToken result: <token>
vso.packaging_write","TargetAccounts":["<id>"],"Token":"[Redacted]","AlternateToken":null,"IsValid":true,"IsPublic":false,"PublicData":null,"Source":null}
CredentialProvider.VSS: Acquired JWT token valid from 24-8-2018 05:32:12 to 24-8-2018 09:42:08 (04:09:56 total, 03:59:56.6725072 remaining). Note that the server may reject the token before the end date due to revocation, etc.
CredentialProvider.VSS: Exit message: 
CredentialProvider.VSS: Exit code: Success
  GET https://account.pkgs.visualstudio.com/_packaging/feed-id/nuget/v2/$metadata
  OK https://account.pkgs.visualstudio.com/_packaging/feed-id/nuget/v2/$metadata 56ms
  GET https://account.pkgs.visualstudio.com/_packaging/feed-id/nuget/v2/Search()?$filter=IsLatestVersion&$orderby=Id&searchTerm='package-id'&targetFramework=''&includePrerelease=false&$skip=0&$top=30&semVerLevel=2.0.0
  OK https://account.pkgs.visualstudio.com/_packaging/feed-id/nuget/v2/Search()?$filter=IsLatestVersion&$orderby=Id&searchTerm='package-id'&targetFramework=''&includePrerelease=false&$skip=0&$top=30&semVerLevel=2.0.0 77ms
package-id
 0.1.11
 other-package-id

**Comment by oskarm93** [comment]
I have exactly the same exception when trying to generate tokens!
A portion of our restores fails with this error:
```
Microsoft.VisualStudio.Services.Common.VssUnauthorizedException: VS30063: You are not authorized to access https://<VSTS_ACCOUNT_NAME>.vssps.visualstudio.com.
```

Raised it previously on Developer Community to no avail:
https://developercommunity.visualstudio.com/content/problem/310187/sporadic-nuget-restore-failures-from-vsts-feed.html

Using the same version of NuGet and VSS Cred Provider from VSTS feed tab, but on Windows Server 2016.

**Comment by nkolev92** [comment]
The VSS Credential provider is not managed by the NuGet itself. It's on the provider to do caching when appropriate. 

Part of the difference you are seeing between having a PAT in the config and an exe could be the process startup cost. 
It's likely that it adds at least 1s to the total. 

Do you get prompted by the credential provider too often? Or is it just the work in the background?

//cc @keithrob @alexmullans

Please note that there'll be a new credential provider by VSTS based on https://docs.microsoft.com/en-us/nuget/reference/extensibility/nuget-cross-platform-authentication-plugin

**Comment by sanderaernouts** [comment]
We do not get prompted often but it does retrieve a new token every time so the provider seems to add more like 10s instead of 1s to every call.

**Comment by jotaylo** [comment]
It looks like the thread on the developer community didn't get assigned to the correct team. I've assigned it now, and we'll follow up over there.

**Comment by nkolev92** [comment]
Closing this as a duplicate of the developer community issue  https://developercommunity.visualstudio.com/content/problem/310187/sporadic-nuget-restore-failures-from-vsts-feed.html per @jotaylo 's comment.

**Comment by jotaylo** [comment]
@sanderaernouts requesting a new token on every invocation to nuget.exe is expected, however we suspect that there has recently been a server-side change that's degrading the performance. We'll see if we can add some sort of optimization here.
 In the meantime, you may find the [Vss NuGet AuthHelper](https://www.nuget.org/packages/Microsoft.VisualStudio.Services.NuGet.AuthHelper/) useful. It lets you supply a NuGet config file, and will generate tokens for the sources defined in that config file and add them to the user-level (global) NuGet config file.

**Comment by sanderaernouts** [comment]
@jotaylo thanks for sharing the AuthHelper. I will certainly have a look at that. Regarding the tokens, I see the token generated by the Credential Provider has an expiration time that is valid for about 4 hours after it was generated. Can't this token be cached/stored securely and be reused until it expires?

We use a very similar approach in one of our applications which stores a generated JWT token using [Windows Data Protection](https://msdn.microsoft.com/en-us/library/ms995355.aspx).

**Comment by irontoby** [comment]
@jotaylo can you provide more info on the VSS NuGet AuthHelper? I can't find any info on it aside from the NuGet site.

I'm currently also using the `CredentialProvider.VSS.exe` and it hangs for 10-15 seconds on each call while fetching the token from VSTS (Azure Devops), so if the AuthHelper can remove that I'd like to give it a try.

**Comment by infin8x** [comment]
The auth helper is quite old and is only recommended for a few very specific scenarios. [artifacts-credprovider](https://github.com/Microsoft/artifacts-credprovider) is the way to go - see if that improves performance for you? Note that you'll need NuGet 4.8.0 or greater.

**Comment by irontoby** [comment]
@alexmullans thanks! I'm running into lots of conflicting info. I just created a Personal Access Token and added it to NuGet using the [NuGet 2.x instructions](https://docs.microsoft.com/en-us/azure/devops/artifacts/nuget/nuget-exe?view=vsts&tabs=new-nav#add-a-feed-to-nuget-2) (which still work for 4.x) and that has dramatically sped up response times. The CredentialProvider you linked though does seem more comprehensive so I'll give that a try.

**Comment by infin8x** [comment]
@irontoby yep, that still works. If you have time, would you mind linking to some of the conflicting info you've encountered? We're planning to streamline some of our docs and product experiences around authentication in Q2 - it'd be helpful to understand what you've found today to ensure we're thinking about it.

**Comment by irontoby** [comment]
@alexmullans maybe not _conflicting_ info but it was pretty difficult figuring out how to connect our Azure package feed in a performant way and to search for additional information on how to setup.

* The "official" setup (i.e. the one offered from the Azure DevOps "Connect to feed" link as [described here](https://docs.microsoft.com/en-us/azure/devops/artifacts/get-started-nuget)) to "Download NuGet + Credential Provider" is just a zipfile with the two `.exe`s and a couple license docs. No instructions whatsoever on where to put `nuget.exe` and the credential provider. Just putting `nuget.exe` someplace in my path was my default guess but had no idea what to do with the credential provider.
* I eventually found [this document](https://docs.microsoft.com/en-us/nuget/reference/extensibility/nuget-exe-credential-providers) & figured out just putting the credential provider in the same directory as `nuget.exe` worked. In hindsight this may seem obvious but when running NuGet from command line, you're in the directory where your dev files are, not where `nuget.exe` is located, so it wasn't obvious where it came into play.
* I then had the problem mentioned in this issue: every NuGet call which contacted Azure seemed to hang for about 10+ seconds. Eventually found the verbose switch which gave me the same exceptions mentioned above & led me here.
* Found the link above for "Vss NuGet AuthHelper" & spent a while trying to get that to work. IIRC it had the same problems as the one downloaded from the zipfile.
* Next found the "artifacts-credprovider" link & spent a while trying to get that to work. I never did get it to even appear to work; `nuget.exe` kept prompting me for credentials & I never got the interactive Azure DevOps login.
* Eventually found the Personal Access Token method [mentioned here](https://docs.microsoft.com/en-us/azure/devops/artifacts/nuget/nuget-exe), which I initially didn't try because it was listed only for NuGet 2.

In total I spent at least half a day trying to get NuGet connecting properly to Azure, which was quite frustrating for something that seemed like it should "just work" considering I was using the full Microsoft offering.

A few more things adding to my frustration:

* The official credential provider still has "VSS" in the title which I assume refers to Visual SourceSafe & caused me to guess that I was working with some abandoned 20-year-old tool.
* The credential provider also keeps showing the "Getting authentication token" message in the verbose logs as mentioned by the OP. Why is it not caching the token somewhere, and if it is then why does it still show that message?
* The "Azure DevOps" name is still new enough that searching for good info on it is still a bit tricky.

**Comment by infin8x** [comment]
@irontoby this is really great, thanks for taking the time to put it together. I've filed it in the feature that tracks improving these workflows, and will refer back to it closer to Q2 as we figure out the right improvements to prioritize.
